### PR TITLE
Protocol v2 changes

### PR DESCRIPTION
**Pull request by singhsarab**
1. Enable use of new Message type.
2. Protocol version check between runner and host
3. Object model changes for reducing the verbosity.
4. Serializers for supported protocol versions.
5. Unit tests

Related to #396.

### CODE REVIEW

File: src/Microsoft.TestPlatform.CommunicationUtilities/Interfaces/ITestRequestSender.cs [hunk 1]
```diff
@@ -21,6 +21,12 @@ public interface ITestRequestSender : IDisposable
         int InitializeCommunication();
 
         /// <summary>
+        /// Used for protocol version check with TestHost
+        /// </summary>
+        /// <returns>true if the version check is successful</returns>
+        bool CheckVersionWithTestHost();
```
Comment (AbhitejJohn) [inline src/Microsoft.TestPlatform.CommunicationUtilities/Interfaces/ITestRequestSender.cs]: would this be used for data collector as well?

File: src/Microsoft.TestPlatform.CommunicationUtilities/JsonDataSerializer.cs [hunk 2]
```diff
@@ -74,28 +82,24 @@ public T DeserializePayload<T>(Message message)
...
-            if (MessageType.TestMessage.Equals(message.MessageType))
-            {
-                retValue = message.Payload.ToObject<T>();
-            }
-            else
-            {
-                retValue = message.Payload.ToObject<T>(serializer);
-            }
+            var versionedMessage = message as VersionedMessage;
+            var serializer = versionedMessage?.Version == 2 ? payloadSerializer2 : payloadSerializer;
```
Comment (AbhitejJohn) [inline src/Microsoft.TestPlatform.CommunicationUtilities/JsonDataSerializer.cs]: can we have a `GetPayloadSerializer(Constants.CurrentVersion)` for this? This can then be reused everywhere here.

File: src/Microsoft.TestPlatform.CommunicationUtilities/JsonDataSerializer.cs [hunk 3]
```diff
@@ -74,28 +82,24 @@ public T DeserializePayload<T>(Message message)
         {
             T retValue = default(T);
 
-            // TODO: Currently we use json serializer auto only for non-testmessage types
-            // CHECK: Can't we just use auto for everything
-            if (MessageType.TestMessage.Equals(message.MessageType))
-            {
-                retValue = message.Payload.ToObject<T>();
-            }
```
Comment (AbhitejJohn) [inline src/Microsoft.TestPlatform.CommunicationUtilities/JsonDataSerializer.cs]: Assuming `MessageType.TestMessage` would continue to work post this change.

File: src/Microsoft.TestPlatform.CommunicationUtilities/JsonDataSerializer.cs [hunk 4]
```diff
@@ -121,30 +125,41 @@ public string SerializeMessage(string messageType)
...
+
+        /// <summary>
         /// Serialize an object to JSON using default serialization settings.
         /// </summary>
         /// <typeparam name="T">Type of object to serialize.</typeparam>
         /// <param name="data">Instance of the object to serialize.</param>
+        /// <param name="version">Version to be stamped.</param>
         /// <returns>JSON string.</returns>
-        public string Serialize<T>(T data)
+        public string Serialize<T>(T data, int version = 1)
```
Comment (AbhitejJohn) [inline src/Microsoft.TestPlatform.CommunicationUtilities/JsonDataSerializer.cs]: shouldn't this by default be the highest?

File: src/Microsoft.TestPlatform.CommunicationUtilities/TestRequestSender.cs [hunk 5]
```diff
@@ -28,18 +28,29 @@ public sealed class TestRequestSender : ITestRequestSender
 
         private IDataSerializer dataSerializer;
 
+        // TODO:sasin Change the version to 2
```
Comment (AbhitejJohn) [inline src/Microsoft.TestPlatform.CommunicationUtilities/TestRequestSender.cs]: when would this change to 2?

File: src/Microsoft.TestPlatform.CommunicationUtilities/TestRequestSender.cs [hunk 6]
```diff
@@ -76,23 +87,51 @@ public void Close()
         }
 
         /// <inheritdoc/>
+        public bool CheckVersionWithTestHost()
```
Comment (AbhitejJohn) [inline src/Microsoft.TestPlatform.CommunicationUtilities/TestRequestSender.cs]: This reminds me - These RFCs([1](https://github.com/Microsoft/vstest-docs/blob/master/RFCs/0003-Test-Execution-Protocol.md) ,[2](https://github.com/Microsoft/vstest-docs/blob/master/RFCs/0002-Test-Discovery-Protocol.md)) are now out of date. Should those change?

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyOperationManager.cs [hunk 7]
```diff
@@ -114,22 +116,40 @@ public virtual void SetupChannel(IEnumerable<string> sources)
...
+                    }
 
-                if (!string.IsNullOrWhiteSpace(this.testHostProcessStdError))
+                    throw new TestPlatformException(string.Format(CultureInfo.CurrentUICulture, errorMsg));
+                }
+
+                // Handling special case for dotnet core projects with older test hosts
+                // Older test hosts are not aware of protocol version check
+                // Hence we should not be sending VersionCheck message to these test hosts
+                bool checkRequired = true;
```
Comment (AbhitejJohn) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyOperationManager.cs]: Is it the ProxyOperationManagers responsibility to perform this check? Or should this be part of an `Initialize` in RequestSender?

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyOperationManager.cs [hunk 8]
```diff
@@ -114,22 +116,40 @@ public virtual void SetupChannel(IEnumerable<string> sources)
...
+                var property = this.testHostManager.GetType().GetRuntimeProperties().FirstOrDefault(p => string.Equals(p.Name, versionCheckPropertyName, StringComparison.OrdinalIgnoreCase));
+                if (property != null)
```
Comment (AbhitejJohn) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyOperationManager.cs]: Don't really know if there is something better we can do here. Seems odd :(
Maybe have a config file with the version and check for its existence?

File: src/Microsoft.TestPlatform.ObjectModel/TestObject.cs [hunk 9]
```diff
@@ -29,6 +29,11 @@ public abstract class TestObject
         private readonly Dictionary<TestProperty, object> store;
 
         /// <summary>
+        /// The store for all the local properties registered.
+        /// </summary>
+        private readonly Dictionary<TestProperty, object> localStore;
```
Comment (AbhitejJohn) [inline src/Microsoft.TestPlatform.ObjectModel/TestObject.cs]: 👍

Comment (AbhitejJohn) [inline src/Microsoft.TestPlatform.ObjectModel/TestObject.cs]: So, this wont be a breaking change to extensions right?

File: src/Microsoft.TestPlatform.ObjectModel/TestResult.cs [hunk 10]
```diff
@@ -383,12 +383,7 @@ public static class TestResultProperties
         [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Security", "CA2104:DoNotDeclareReadOnlyMutableReferenceTypes")]
         public static readonly TestProperty ErrorStackTrace = TestProperty.Register("TestResult.ErrorStackTrace", Resources.Resources.TestResultPropertyErrorStackTraceLabel, typeof(string), typeof(TestResult));
 #endif
-
-        private static bool ValidateComputerName(object value)
```
Comment (AbhitejJohn) [inline src/Microsoft.TestPlatform.ObjectModel/TestResult.cs]: We don't like this? Why?

File: src/Microsoft.TestPlatform.CommunicationUtilities/Interfaces/ITestRequestSender.cs (hunk 1 above)
Comment (singhsarab) [inline src/Microsoft.TestPlatform.CommunicationUtilities/Interfaces/ITestRequestSender.cs]: Yes, ideally we should. Will take it up separately.

File: src/Microsoft.TestPlatform.CommunicationUtilities/TestRequestSender.cs (hunk 5 above)
Comment (singhsarab) [inline src/Microsoft.TestPlatform.CommunicationUtilities/TestRequestSender.cs]: Most likely in next PR, once IDE to runner is complete.

File: src/Microsoft.TestPlatform.ObjectModel/TestObject.cs (hunk 9 above)
Comment (singhsarab) [inline src/Microsoft.TestPlatform.ObjectModel/TestObject.cs]: No, it wont.

**Review by AbhitejJohn — APPROVED** [review]

File: src/Microsoft.TestPlatform.CommunicationUtilities/Interfaces/ITestRequestSender.cs (hunk 1 above)
Comment (codito) [inline src/Microsoft.TestPlatform.CommunicationUtilities/Interfaces/ITestRequestSender.cs]: Datacollector will always be bundled with the test runner. So they can communicate in any protocol without any negotiation?

File: src/Microsoft.TestPlatform.CommunicationUtilities/JsonDataSerializer.cs [hunk 11]
```diff
@@ -20,26 +19,33 @@ public class JsonDataSerializer : IDataSerializer
...
+                DateParseHandling = DateParseHandling.DateTimeOffset,
+                DateTimeZoneHandling = DateTimeZoneHandling.Utc,
+                TypeNameHandling = TypeNameHandling.None
+            };
+
+            payloadSerializer = JsonSerializer.Create(jsonSettings);
+            payloadSerializer2 = JsonSerializer.Create(jsonSettings);
+
+            payloadSerializer.ContractResolver = new TestPlatformContractResolver1();
+            payloadSerializer2.ContractResolver = new DefaultTestPlatformContractResolver();
```
Comment (codito) [inline src/Microsoft.TestPlatform.CommunicationUtilities/JsonDataSerializer.cs]: When v3 comes in, we need to change this code? Why not consider versioning the ContractResolvers suffixed with version number. The class name need not say what's default, let it be determined only in this file.

File: src/Microsoft.TestPlatform.CommunicationUtilities/JsonDataSerializer.cs [hunk 12]
```diff
@@ -74,28 +82,24 @@ public T DeserializePayload<T>(Message message)
...
-            // CHECK: Can't we just use auto for everything
-            if (MessageType.TestMessage.Equals(message.MessageType))
-            {
-                retValue = message.Payload.ToObject<T>();
-            }
-            else
-            {
-                retValue = message.Payload.ToObject<T>(serializer);
-            }
+            var versionedMessage = message as VersionedMessage;
```
Comment (codito) [inline src/Microsoft.TestPlatform.CommunicationUtilities/JsonDataSerializer.cs]: Please log the default serializer determined here.

File: src/Microsoft.TestPlatform.CommunicationUtilities/JsonDataSerializer.cs [hunk 13]
```diff
@@ -74,28 +82,24 @@ public T DeserializePayload<T>(Message message)
...
 
         /// <summary>
         /// Deserialize raw JSON to an object using the default serializer.
         /// </summary>
         /// <param name="json">JSON string.</param>
+        /// <param name="version">Version of serializer to be used.</param>
         /// <typeparam name="T">Target type to deserialize.</typeparam>
         /// <returns>An instance of <see cref="T"/>.</returns>
-        public T Deserialize<T>(string json)
+        public T Deserialize<T>(string json, int version = 1)
```
Comment (codito) [inline src/Microsoft.TestPlatform.CommunicationUtilities/JsonDataSerializer.cs]: Should the caller be aware of `version` of payload? How would it know, since the version is embedded inside the json?

File: src/Microsoft.TestPlatform.CommunicationUtilities/JsonDataSerializer.cs [hunk 14]
```diff
@@ -121,30 +125,41 @@ public string SerializeMessage(string messageType)
...
-            // CHECK: Can't we just use auto for everything
-            if (MessageType.TestMessage.Equals(messageType))
-            {
-                serializedPayload = JToken.FromObject(payload);
-            }
-            else
-            {
-                serializedPayload = JToken.FromObject(payload, serializer);
-            }
+            var serializedPayload = JToken.FromObject(payload, payloadSerializer);
```
Comment (codito) [inline src/Microsoft.TestPlatform.CommunicationUtilities/JsonDataSerializer.cs]: Why are these APIs not using versioned messages? How would the caller decide which API to use (and when)?

File: src/Microsoft.TestPlatform.CommunicationUtilities/Messages/VersionedMessage.cs [hunk 15]
```diff
@@ -0,0 +1,16 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.VisualStudio.TestPlatform.CommunicationUtilities
+{
+    /// <summary>
+    /// New construct with version used for communication
```
Comment (codito) [inline src/Microsoft.TestPlatform.CommunicationUtilities/Messages/VersionedMessage.cs]: `New` becomes old ;) Better to not keep a sense of time in documentation. Add a remark specifying which version of vstest this was introduced and that this is the default message for protocol 2+.

File: test/Microsoft.TestPlatform.CommunicationUtilities.UnitTests/Serialization/TestCaseSerializationTests.cs [hunk 16]
```diff
@@ -105,44 +108,120 @@ public void TestCaseObjectShouldDeserializeEscapedWindowsPath()
...
+            var test = Deserialize<TestCase>(json, 2);
+
+            Assert.AreEqual(@"C:\Test\TestAssembly.dll", test.Source);
+        }
+
+        #endregion
+
+        #region Common Tests
+
         [TestMethod]
```
Comment (codito) [inline test/Microsoft.TestPlatform.CommunicationUtilities.UnitTests/Serialization/TestCaseSerializationTests.cs]: `DataTestMethod`? Same for below test too.

File: src/Microsoft.TestPlatform.CommunicationUtilities/Serialization/TestCaseConverter.cs [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.VisualStudio.TestPlatform.CommunicationUtilities.Serialization
+{
+    using System;
+    using Microsoft.VisualStudio.TestPlatform.ObjectModel;
+    using Newtonsoft.Json;
+
+    /// <inheritdoc/>
+    public class TestCaseConverter : JsonConverter
```
Comment (codito) [inline src/Microsoft.TestPlatform.CommunicationUtilities/Serialization/TestCaseConverter.cs]: Not clear when this is used, why is it required. Require documentation.

File: src/Microsoft.TestPlatform.CommunicationUtilities/Serialization/TestPlatformContractResolver1.cs [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+        /// <inheritdoc/>
+        protected override JsonContract CreateContract(Type objectType)
+        {
+            var contract = base.CreateContract(objectType);
+            if (typeof(TestCase) == objectType)
+            {
+                contract.Converter = new TestCaseConverter();
+            }
+
+            if (typeof(TestResult) == objectType)
```
Comment (codito) [inline src/Microsoft.TestPlatform.CommunicationUtilities/Serialization/TestPlatformContractResolver1.cs]: `if` always executes even if above comparison is true.

File: src/Microsoft.TestPlatform.CommunicationUtilities/Serialization/TestResultConverter.cs [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.VisualStudio.TestPlatform.CommunicationUtilities.Serialization
+{
+    using System;
+    using Microsoft.VisualStudio.TestPlatform.ObjectModel;
+    using Newtonsoft.Json;
+
+    /// <inheritdoc/>
+    public class TestResultConverter : JsonConverter
```
Comment (codito) [inline src/Microsoft.TestPlatform.CommunicationUtilities/Serialization/TestResultConverter.cs]: Require documentation. Should we include these classes in versioned folders (same for TestCaseConverter)?

File: test/Microsoft.TestPlatform.CommunicationUtilities.UnitTests/Serialization/TestCaseSerializationTests.cs [hunk 20]
```diff
@@ -28,6 +28,8 @@ public class TestCaseSerializationTests
                                                    Traits = { new Trait("Priority", "0"), new Trait("Category", "unit") }
                                                };
 
+        #region v1 Tests
```
Comment (codito) [inline test/Microsoft.TestPlatform.CommunicationUtilities.UnitTests/Serialization/TestCaseSerializationTests.cs]: Require test where a TestCase doesn't have any properties and can be serialized properly.

File: src/Microsoft.TestPlatform.CommunicationUtilities/Serialization/TestCaseConverter.cs [hunk 21]
```diff
@@ -0,0 +1,47 @@
...
+        /// <inheritdoc/>
+        public override bool CanConvert(Type objectType)
+        {
+            return typeof(TestCase) == objectType;
+        }
+
+        /// <inheritdoc/>
+        public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
+        {
+            throw new NotImplementedException();
```
Comment (codito) [inline src/Microsoft.TestPlatform.CommunicationUtilities/Serialization/TestCaseConverter.cs]: Why are we not using this for deserialization? Please add a note.

File: src/Microsoft.TestPlatform.CommunicationUtilities/JsonDataSerializer.cs [hunk 22]
```diff
@@ -121,30 +125,41 @@ public string SerializeMessage(string messageType)
...
         }
 
         /// <summary>
+        /// Serialize a message with payload.
+        /// </summary>
+        /// <param name="messageType">Type of the message.</param>
+        /// <param name="payload">Payload for the message.</param>
+        /// <param name="version">Version for the message.</param>
+        /// <returns>Serialized message.</returns>
+        public string SerializePayload(string messageType, object payload, int version)
```
Comment (codito) [inline src/Microsoft.TestPlatform.CommunicationUtilities/JsonDataSerializer.cs]: How is this tested?

File: test/Microsoft.TestPlatform.CommunicationUtilities.UnitTests/Serialization/TestResultSerializationTests.cs [hunk 23]
```diff
@@ -58,15 +61,15 @@ public void TestResultJsonShouldContainAllPropertiesOnSerialization()
...
...d\":\"TestResult.ErrorStackTrace\",\"Label\":\"Error Stack Trace\",\"Category\":\"\",\"Description\":\"\",\"Attributes\":0,\"ValueType\":\"System.String\"},\"Value\":\"sampleStackTrace\"},{\"Key\":{\"Id\":\"TestResult.DisplayName\",\"Label\":\"TestResult Display Name\",\"Category\":\"\",\"Description\":\"\",\"Attributes\":1,\"ValueType\":\"System.String\"},\"Value\":\"sampleTestResult\"},{\"Key\":{\"Id\":\"TestResult.ComputerName\",\"Label\":\"Computer Name\",\"Category\":\"\",\"Description\":\"\",\"Attributes\":0,\"ValueType\":\"System.String\"},\"Value\":\"sampleComputerName\"},{\"Key\":{\"Id\":\"TestResult.Duration\",\"Label\":\"Duration\",\"Category\":\"\",\"Description\":\"\",\"Attributes\":0,\"ValueType\":\"System.TimeSpan\"},\"Value\":\"10675199.02:48:05.4775807\"},{\"Key\":{\"Id\":\"TestResult.StartTime\",\"Label\":\"Start Time\",\"Category\":\"\",\"Description\":\"\",\"Attributes\":0,\"ValueType\":\"System.DateTimeOffset\"},\"Value\":\"2007-03-10T00:00:00+00:00\"},{\"Key\":{\"Id\":\"TestResult.EndTime\",\"Label\":\"End Time\",\"Category\":\"\",\"Description\":\"\",\"Attributes\":0,\"ValueType\":\"System.DateTimeOffset\"},\"Value\":\"9999-12-31T23:59:59.9999999+00:00\"}]}";
```
Comment (codito) [inline test/Microsoft.TestPlatform.CommunicationUtilities.UnitTests/Serialization/TestResultSerializationTests.cs]: Why is this change required?

File: test/Microsoft.TestPlatform.CommunicationUtilities.UnitTests/Serialization/TestResultSerializationTests.cs [hunk 24]
```diff
@@ -90,7 +93,6 @@ public void TestResultObjectShouldSerializeAttachments()
             var result = new TestResult(testCase);
             result.Attachments.Add(new AttachmentSet(new Uri("http://dummyUri"), "sampleAttachment"));
             var expectedJson = "{\"TestCase\":{\"Properties\":[{\"Key\":{\"Id\":\"TestCase.FullyQualifiedName\",\"Label\":\"FullyQualifiedName\",\"Category\":\"\",\"Description\":\"\",\"Attributes\":1,\"ValueType\":\"System.String\"},\"Value\":\"sampleTestClass.sampleTestCase\"},{\"Key\":{\"Id\":\"TestCase.ExecutorUri\",\"Label\":\"Executor Uri\",\"Category\":\"\",\"Description\":\"\",\"Attributes\":1,\"ValueType\":\"System.Uri\"},\"Value\":\"executor://sampleTestExecutor\"},{\"Key\":{\"Id\":\"TestCase.Source\",\"Label\":\"Source\",\"Category\":\"\",\"Description\":\"\",\"Attributes\":0,\"ValueType\":\"System.String\"},\"Value\":\"sampleTest.dll\"}]},\"Attachments\":[{\"Uri\":\"http://dummyUri\",\"DisplayName\":\"sampleAttachment\",\"Attachments\":[]}],\"Messages\":[],\"Properties\":[]}";
-
```
Comment (codito) [inline test/Microsoft.TestPlatform.CommunicationUtilities.UnitTests/Serialization/TestResultSerializationTests.cs]: Please undo. New line is required between `Arrange` and `Act` parts of a test method.

File: test/Microsoft.TestPlatform.CommunicationUtilities.UnitTests/Serialization/TestResultSerializationTests.cs [hunk 25]
```diff
@@ -108,14 +110,84 @@ public void TestResultObjectShouldDeserializeAttachments()
...
+            Assert.AreEqual(testResult.StartTime, test.StartTime);
+        }
+
+        [TestMethod]
+        public void TestResultObjectShouldSerializeAttachmentsV2()
+        {
+            var result = new TestResult(testCase);
+            result.Attachments.Add(new AttachmentSet(new Uri("http://dummyUri"), "sampleAttachment"));
+            var expectedJson = "{\"TestCase\":{\"Properties\":[{\"Key\":{\"Id\":\"TestCase.FullyQualifiedName\",\"Label\":\"FullyQualifiedName\",\"Category\":\"\",\"Description\":\"\",\"Attributes\":1,\"ValueType\":\"System.String\"},\"Value\":\"sampleTestClass.sampleTestCase\"},{\"Key\":{\"Id\":\"TestCase.ExecutorUri\",\"Label\":\"Executor Uri\",\"Category\":\"\",\"Description\":\"\",\"Attributes\":1,\"ValueType\":\"System.Uri\"},\"Value\":\"executor://sampleTestExecutor\"},{\"Key\":{\"Id\":\"TestCase.Source\",\"Label\":\"Source\",\"Category\":\"\",\"Description\":\"\",\"Attributes\":0,\"ValueType\":\"System.String\"},\"Value\":\"sampleTest.dll\"}]},\"Attachments\":[{\"Uri\":\"http://dummyUri\",\"DisplayName\":\"sampleAttachment\",\"Attachments\":[]}],\"Messages\":[],\"Properties\":[]}";
+            var json = Serialize(result);
```
Comment (codito) [inline test/Microsoft.TestPlatform.CommunicationUtilities.UnitTests/Serialization/TestResultSerializationTests.cs]: New line between arrange and act. Fix elsewhere in file.

File: src/Microsoft.TestPlatform.CommunicationUtilities/SocketCommunicationManager.cs [hunk 26]
```diff
@@ -216,6 +216,18 @@ public void SendMessage(string messageType, object payload)
         }
 
         /// <summary>
+        ///  Writes message to the binary writer with payload
+        /// </summary>
+        /// <param name="messageType">Type of Message to be sent, for instance TestSessionStart</param>
+        /// <param name="payload">payload to be sent</param>
+        /// <param name="version">version to be sent</param>
+        public void SendMessage(string messageType, object payload, int version)
```
Comment (codito) [inline src/Microsoft.TestPlatform.CommunicationUtilities/SocketCommunicationManager.cs]: Require tests?

File: src/Microsoft.TestPlatform.CommunicationUtilities/TestRequestSender.cs [hunk 27]
```diff
@@ -76,23 +87,51 @@ public void Close()
...
+            {
+                var protocolVersion = this.dataSerializer.DeserializePayload<int>(message);
+
+                // TODO: Should we check if this is valid ?
+                this.highestNegotiatedVersion = protocolVersion;
+                success = true;
+            }
+            else if (message.MessageType == MessageType.ProtocolError)
+            {
+                EqtTrace.Error("TestRequestSender: VersionCheck Failed");
```
Comment (codito) [inline src/Microsoft.TestPlatform.CommunicationUtilities/TestRequestSender.cs]: Protocol.Error doesn't specify what's the highest version supported at the other end? How does user find why version check failed?

File: src/Microsoft.TestPlatform.CommunicationUtilities/TestRequestSender.cs [hunk 28]
```diff
@@ -76,23 +87,51 @@ public void Close()
...
+            this.communicationManager.SendMessage(MessageType.VersionCheck, payload: this.highestNegotiatedVersion);
+
+            var message = this.communicationManager.ReceiveMessage();
+
+            if (message.MessageType == MessageType.VersionCheck)
+            {
+                var protocolVersion = this.dataSerializer.DeserializePayload<int>(message);
+
+                // TODO: Should we check if this is valid ?
+                this.highestNegotiatedVersion = protocolVersion;
```
Comment (codito) [inline src/Microsoft.TestPlatform.CommunicationUtilities/TestRequestSender.cs]: Log this please. What does `TODO` mean?

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyOperationManager.cs (hunk 7 above)
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyOperationManager.cs]: Can this move out as an static extension method to `ITestHostManager`?

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyOperationManager.cs [hunk 29]
```diff
@@ -114,22 +116,40 @@ public virtual void SetupChannel(IEnumerable<string> sources)
...
+                    checkRequired = (bool)property.GetValue(this.testHostManager);
+                }
+
+                if (checkRequired)
                 {
-                    // Testhost failed with error
-                    errorMsg = string.Format(CrossPlatEngineResources.TestHostExitedWithError, this.testHostProcessStdError);
+                    if (!this.RequestSender.CheckVersionWithTestHost())
+                    {
+                        throw new TestPlatformException(string.Format(CultureInfo.CurrentUICulture, "Protocol version check failed"));
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyOperationManager.cs]: Add to resource.

File: src/Microsoft.TestPlatform.CrossPlatEngine/EventHandlers/TestRequestHandler.cs [hunk 30]
```diff
@@ -212,20 +231,20 @@ public void Close()
         /// <inheritdoc/>
         public void SendTestCases(IEnumerable<TestCase> discoveredTestCases)
         {
-            this.communicationManager.SendMessage(MessageType.TestCasesFound, discoveredTestCases);
+            this.communicationManager.SendMessage(MessageType.TestCasesFound, discoveredTestCases, protocolVersion);
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/EventHandlers/TestRequestHandler.cs]: `this.protocolVersion`

File: test/Microsoft.TestPlatform.CommunicationUtilities.UnitTests/TestRequestSenderTests.cs [hunk 31]
```diff
@@ -72,12 +73,42 @@ public void DisposeShouldCallStopServerOnCommunicationManager()
         }
 
         [TestMethod]
+        public void VersionCheckWithTestHostShouldCheckVersionIfVersionCheckPassesReturnTrue()
```
Comment (codito) [inline test/Microsoft.TestPlatform.CommunicationUtilities.UnitTests/TestRequestSenderTests.cs]: s/IfVersionCheck/IsVersionCheck

Comment (codito) [inline test/Microsoft.TestPlatform.CommunicationUtilities.UnitTests/TestRequestSenderTests.cs]: Or may be we are missing a test for `IsVersionCheckRequired` property?

File: test/Microsoft.TestPlatform.CommunicationUtilities.UnitTests/TestRequestSenderTests.cs [hunk 32]
```diff
@@ -277,7 +308,7 @@ public void StartTestRunWithSourcesShouldCallHandleTestRunStatsChange()
 
             waitHandle.WaitOne();
 
-            this.mockCommunicationManager.Verify(mc => mc.SendMessage(MessageType.StartTestExecutionWithSources, runCriteria), Times.Once);
+            this.mockCommunicationManager.Verify(mc => mc.SendMessage(MessageType.StartTestExecutionWithSources, runCriteria, this.version), Times.Once);
```
Comment (codito) [inline test/Microsoft.TestPlatform.CommunicationUtilities.UnitTests/TestRequestSenderTests.cs]: Better to extract a method `VerifySendMessage(messageType, payload)` for all repetitions.

File: src/Microsoft.TestPlatform.CrossPlatEngine/EventHandlers/TestRequestHandler.cs [hunk 33]
```diff
@@ -80,8 +86,21 @@ public void ProcessRequests(ITestHostManagerFactory testHostManagerFactory)
...
                 {
+                    case MessageType.VersionCheck:
+                        var version = this.dataSerializer.DeserializePayload<int>(message);
+                        this.protocolVersion = Math.Min(version, highestSupportedVersion);
+                        this.communicationManager.SendMessage(MessageType.VersionCheck, this.protocolVersion);
+
+                        // Can only do this after InitializeCommunication because TestHost cannot "Send Log" unless communications are initialized
+                        if (!string.IsNullOrEmpty(EqtTrace.LogFile))
+                        {
+                            this.SendLog(TestMessageLevel.Informational, string.Format("Logging TestHost Diagnostics in file: {0}", EqtTrace.LogFile));
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/EventHandlers/TestRequestHandler.cs]: Should this be added to a resource?

File: src/Microsoft.TestPlatform.ObjectModel/TestObject.cs [hunk 34]
```diff
@@ -269,7 +295,7 @@ private void PrivateSetPropertyValue(TestProperty property, object value)
         /// Convert passed in value from TestProperty's specified value type.
         /// </summary>
         /// <returns>Converted object</returns>
-        private static object ConvertPropertyFrom<T>(TestProperty property, CultureInfo culture, object value)
+        internal static object ConvertPropertyFrom<T>(TestProperty property, CultureInfo culture, object value)
```
Comment (codito) [inline src/Microsoft.TestPlatform.ObjectModel/TestObject.cs]: Why is this required?

File: src/Microsoft.TestPlatform.ObjectModel/TestObject.cs [hunk 35]
```diff
@@ -224,7 +239,18 @@ public void SetPropertyValue<T>(TestProperty property, LazyPropertyValue<T> valu
...
             object objValue = ConvertPropertyFrom<T>(property, culture, value);
 
-            this.PrivateSetPropertyValue(property, objValue);
+            this.PrivateSetPropertyValue(property, objValue, this.store);
+        }
+
+        /// <summary>
+        /// Set TestProperty's value to the specified value T.
+        /// </summary>
+        protected void SetLocalPropertyValue<T>(TestProperty property, T value)
```
Comment (codito) [inline src/Microsoft.TestPlatform.ObjectModel/TestObject.cs]: I am not sure if this is the right thing to do. In protocol V2 too, user will unnecessarily pay the cost of an additional dictionary and hashing several properties in it.

File: test/Microsoft.TestPlatform.CommunicationUtilities.UnitTests/Serialization/TestResultSerializationTests.cs (hunk 23 above)
Comment (singhsarab) [inline test/Microsoft.TestPlatform.CommunicationUtilities.UnitTests/Serialization/TestResultSerializationTests.cs]: Changed the testcase to have realistic value rather than default min value.

File: src/Microsoft.TestPlatform.ObjectModel/TestObject.cs (hunk 34 above)
Comment (singhsarab) [inline src/Microsoft.TestPlatform.ObjectModel/TestObject.cs]: Not required. Removed

File: src/Microsoft.TestPlatform.CommunicationUtilities/Serialization/TestResultConverter.cs (hunk 19 above)
Comment (singhsarab) [inline src/Microsoft.TestPlatform.CommunicationUtilities/Serialization/TestResultConverter.cs]: Done.

File: src/Microsoft.TestPlatform.CommunicationUtilities/JsonDataSerializer.cs (hunk 14 above)
Comment (singhsarab) [inline src/Microsoft.TestPlatform.CommunicationUtilities/JsonDataSerializer.cs]: We might want to change the APIs.
This is used for payload serialization/deserialization. Currently, the APIs caller reads the version beforehand

File: src/Microsoft.TestPlatform.CommunicationUtilities/JsonDataSerializer.cs (hunk 13 above)
Comment (singhsarab) [inline src/Microsoft.TestPlatform.CommunicationUtilities/JsonDataSerializer.cs]: This is just the payload deserialization, caller reads the version from the message.
